File: monografia/script/reducer_job2.py
import subprocess
import pandas as pd

nomi_mesi = ["Gen", "Feb", "Mar", "Apr", "Mag", "Giu",
             "Lug", "Ago", "Set", "Ott", "Nov", "Dic"]

nome_file_precedente = subprocess.check_output(['hadoop', 'fs', '-text', '/homework/job1/output/medie_calcolate']).decode('utf-8')
dati_precedenti = pd.read_csv(nome_file_precedente, sep='\t')

env = {}

while True:
    line = input()
    if not line:
        break
    
    valore_linea = line.strip().split("\t")
    vendita = {'data': valore_linea[0], 'importo': float(valore_linea[1])}
    risultato = {'totale': 0, 'count': 0}
    
    if vendita['data'] in env:
        valore_corrente = env[vendita['data']]
        media_del_mese = dati_precedenti[dati_precedenti['data'] == vendita['data']]['media'].values[0]
        risultato['count'] = valore_corrente['count'] + 1
        risultato['totale'] = valore_corrente['totale'] + (vendita['importo'] - dati_precedenti[dati_precedenti['data'] == vendita['data']]['media'].values[0]) ** 2
    else:
        media_del_mese = dati_precedenti[dati_precedenti['data'] == vendita['data']]['media'].values[0]
        risultato['count'] = 1
        risultato['totale'] = (vendita['importo'] - dati_precedenti[dati_precedenti['data'] == vendita['data']]['media'].values[0]) ** 2
    
    env[vendita['data']] = risultato

for data, valore in env.items():
    mese = data[5:7]
    anno = data[0:4]
    mese_testuale = nomi_mesi[int(mese) - 1]
    data_formattata = f"{anno}-{mese_testuale}"
    varianza = round(valore['totale'] / valore['count'], 2)
    print(f"{data}\t{varianza}\t{data_formattata}")

File: monografia/script/reducer_job2.r
library(data.table)
nomi_mesi <- c("Gen", "Feb", "Mar", "Apr", "Mag", "Giu",
               "Lug", "Ago", "Set", "Ott", "Nov", "Dic")

nome_file_precedente <- "hadoop fs -text /homework/job1/output/medie_calcolate"
dati_precedenti <- fread(nome_file_precedente)

env <- new.env(hash = TRUE)
connessione <- file("stdin", open = "r")
while (length(line <- readLines(connessione, n = 1, warn = FALSE)) > 0) {
  
  valore_linea <- unlist(strsplit(line, "\t"))
  vendita <- list(data = valore_linea[1], importo = as.numeric(valore_linea[2]))
  risultato <- list(totale  = 0, count = 0)
  
  if (exists(vendita$data, envir = env, inherits = FALSE)) {
      valore_corrente <- get(vendita$data, envir = env)

      risultato$count <- valore_corrente$count + 1
      media_del_mese <- dati_precedenti[dati_precedenti$V1 == vendita$data]$V2
      risultato$totale <- (valore_corrente$totale + (vendita$importo - media_del_mese)^2)
  } else {
    risultato$count <- 1
    media_del_mese <- dati_precedenti[dati_precedenti$V1 == vendita$data]$V2
    risultato$totale <- ((vendita$importo - media_del_mese)^2)
  }

  assign(vendita$data, risultato, envir = env)
}
close(connessione)

for (data in ls(env, all = TRUE)) {
  valore <- get(data, envir = env)
  mese = substring(data, 5, 6)
  anno = substring(data, 1, 4)
  mese_testuale = nomi_mesi[as.numeric(mese)]
  data_formattata = paste(anno, mese_testuale, sep = "-")
  varianza = round(valore$totale / valore$count, digits = 2)
  cat(data, "\t", varianza ,"\t",data_formattata, "\n", sep = "")
}
